File: C/concorrente_testes.c
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>
#include <time.h>
#include <sys/timeb.h>

#define VIVO 1
#define MORTO 0
#define HIGH 1
#define NORMAL 0

void imprime_matriz(int** matriz, int n){
    for (int i = 0; i < n; i++){
        for (int j = 0; j < n; j++){
            printf("%d ", matriz[i][j]);
        }
        printf("\n");
    }
}

int getNeighbors(int** grid, int i, int j, int n){
    int count = 0;
    int i_local, j_local;
    for (int k = -1; k < 2; k++){
        for (int l = -1; l < 2; l++){
            i_local = i + k;
            j_local = j + l;
            if (i_local < 0){ // infinito pela esquerda
                i_local = n - 1;
            } else if (i_local == n){
                i_local = 0; // infinito pela direita
            }
            if (j_local < 0){
                j_local = n - 1; // infinito por cima
            } else if (j_local == n){
                j_local = 0; // infinito por baixo
            }
            if (grid[i_local][j_local] > 0 && (i_local != i || j_local != j)){
                count++;
            }
        }
    }

    return count;
}

int somaMatriz(int **grid, int n){
    int count = 0;

    for (int i = 0; i < n; i++){
        for (int j = 0; j < n; j++){
            count += grid[i][j];
        }
    }

    return count;
}

void glider(int **grid){
    int lin = 1, col = 1;
    grid[lin+1][col+2] = 1;
    grid[lin+2][col  ] = 1;
    grid[lin  ][col+1] = 1;
    grid[lin+2][col+1] = 1;
    grid[lin+2][col+2] = 1;

}

void pentomino(int **grid){
    int lin = 10, col = 30;
    grid[lin  ][col+1] = 1;
    grid[lin  ][col+2] = 1;
    grid[lin+1][col  ] = 1;
    grid[lin+1][col+1] = 1;
    grid[lin+2][col+1] = 1;

}

void copia_matriz(int **destino, int **origem, int n){
    for (int i = 0; i < n; i++){
        for (int j = 0; j < n; j++){
            destino[i][j] = origem[i][j];
        }
    }
}

int main(){

    int **grid;             //grid atual
    int **newgrid;          //proximo grid
    int N = 2048;           //NxN
    int iteracoes = 2000;    //Numero de iteracoes
    int cont_aux;           //Contador auxiliar
    int tipo = NORMAL;        //Tipo do jogo
    int i, j, k, l, m;         //Auxiliares dos lacos
    int n_teste = 50;            //Numero de execuções
    long diff_laco_atual, diff_paralelo, diff_total;    //Diferenca do tempo em ms
    long media_laco, media_medias_laco, media_paralelos, media_total;
    struct timeb fim_laco, inicio_laco, fim_prog, inicio_prog, fim_paralelo, inicio_paralelo;
    
    for (m = 0; m < n_teste; m++){
        printf("Teste %d\n", m+1);
        media_laco = 0;
        ftime(&inicio_prog);
        //printf("Numero de processadores: %d\n",omp_get_num_procs());

        grid = calloc(N, sizeof(int *));
        newgrid = calloc(N, sizeof(int *));

        for (j = 0; j < N; j++){
            grid[j] = calloc(N,sizeof(int));
            newgrid[j] = calloc(N,sizeof(int));
        }

        glider(grid);
        pentomino(grid);

        //printf("Condicao inicial: %d",somaMatriz(grid,N));

        ftime(&inicio_paralelo);
        for (i = 0; i < iteracoes; i++){
            ftime(&inicio_laco);
            #pragma omp parallel num_threads(1)
                {
                #pragma omp for private(k,l,cont_aux) collapse(2)
                for (k = 0; k < N; k++){
                    for (l = 0; l < N; l++){
                        cont_aux = getNeighbors(grid,k,l,N);
                        if (grid[k][l] == VIVO){
                            if (cont_aux < 2 || cont_aux >= 4){
                                newgrid[k][l] = MORTO;
                            } else {
                                newgrid[k][l] = VIVO;
                            }
                        } else {
                            if (tipo != HIGH){
                                if (cont_aux == 3){
                                    newgrid[k][l] = VIVO;
                                } else {
                                    newgrid[k][l] = MORTO;
                                }
                            } else {
                                if (cont_aux == 3 || cont_aux == 6){
                                    newgrid[k][l] = VIVO;
                                } else {
                                    newgrid[k][l] = MORTO;
                                }
                            }
                        }
                    }
                }
            }
            copia_matriz(grid,newgrid,N);
            //if (i+1 == iteracoes)
            //    printf("\nGeracao %d: %d",i+1,somaMatriz(grid,N));
            ftime(&fim_laco);
            diff_laco_atual = (long) (1000.0 * (fim_laco.time-inicio_laco.time)+(fim_laco.millitm-inicio_laco.millitm));
            if (i == 0)
                media_laco = diff_laco_atual;
            media_laco = (media_laco + diff_laco_atual)/2;
        }
        ftime(&fim_paralelo);  
        ftime(&fim_prog);
        diff_paralelo = (long) (1000.0 * (fim_paralelo.time-inicio_paralelo.time)+(fim_paralelo.millitm-inicio_paralelo.millitm));
        diff_total = (long) (1000.0 * (fim_prog.time-inicio_prog.time)+(fim_prog.millitm-inicio_prog.millitm));
        //printf("\nTempo total: %ums", diff_total);
        //printf("\nMedia dos lacos: %ums",media_laco);
        //printf("\nTempo paralelo: %ums",diff_paralelo);

        if (m == 0) {
            media_medias_laco = media_laco;
            media_paralelos = diff_paralelo;
            media_total = diff_total;
        }

        media_medias_laco = (media_medias_laco + media_laco)/2;
        media_paralelos = (media_paralelos + diff_paralelo)/2;
        media_total = (media_total + diff_total)/2;

        free(grid);
        free(newgrid);
    }

    printf("Media totais: %ums\n",media_total);
    printf("Media paralelos: %ums\n",media_paralelos);
    printf("Media lacos: %ums",media_medias_laco);

    return 0;
}

File: Java/src/ConcorrenteTestes.java
public class ConcorrenteTestes implements Runnable {
    private int[][] grid;
    private int[][] newgrid;
    private int N, inicio_vetor, fim_vetor, tipo;
    private static final int VIVO = 1;
    private static final int MORTO = 0;
    private static final int HIGH = 1;
    private static final int NORMAL = 0;

    public ConcorrenteTestes(int[][] grid_atual, int[][] grid_futuro, int n, int i, int j, int t) {
        N = n;
        inicio_vetor = i;
        fim_vetor = j;
        grid = grid_atual;
        newgrid = grid_futuro;
        tipo = t;
    }

    @Override
    public void run() {
        int cont_aux;
        for (int k = inicio_vetor; k < fim_vetor; k++) {
            for (int l = 0; l < N; l++) {
                cont_aux = Utilitarios.getNeighbors(grid, k, l, N);
                if (grid[k][l] == VIVO) {
                    if (cont_aux < 2 || cont_aux >= 4) {
                        newgrid[k][l] = MORTO;
                    } else {
                        newgrid[k][l] = VIVO;
                    }
                } else {
                    if (tipo != HIGH) {
                        if (cont_aux == 3) {
                            newgrid[k][l] = VIVO;
                        } else {
                            newgrid[k][l] = MORTO;
                        }
                    } else {
                        if (cont_aux == 3 || cont_aux == 6) {
                            newgrid[k][l] = VIVO;
                        } else {
                            newgrid[k][l] = MORTO;
                        }
                    }
                }
            }
        }
    }

    public static void main(String[] args) {
        int N = 2048, iteracoes = 2000, aux;
        int[][] grid = new int[N][N];
        int[][] newgrid = new int[N][N];
        int num_threads = 8;
        int tipo = HIGH;
        long total_inicial, total_final, laco_inicial = 0, laco_final = 0, paralelo_inicial = 0, paralelo_final = 0, media_laco = 0;
        long media_medias_laco = 0, media_total = 0, media_paralelo = 0;
        int n_teste = 5;
        for (int m = 0; m < n_teste; m++) {
            System.out.println("Teste "+(m+1));
            media_laco = 0;
            total_inicial = System.currentTimeMillis();
            Utilitarios.glider(grid);
            Utilitarios.pentomino(grid);
//            System.out.println("Condicao inicial: " + Utilitarios.somaMatriz(grid, N));
            paralelo_inicial = System.currentTimeMillis();
            for (int i = 0; i < iteracoes; i++) {
                laco_inicial = System.currentTimeMillis();
                aux = 0;
                ConcorrenteTestes[] threads = new ConcorrenteTestes[num_threads];
                Thread[] arraythreads = new Thread[num_threads];
                for (int j = 0; j < num_threads; j++) {
                    if (j != num_threads - 1) {
                        threads[j] = new ConcorrenteTestes(grid, newgrid, N, aux, (aux + (N / num_threads)), tipo);
                        aux = (N / num_threads) + 1;
                    } else {
                        threads[j] = new ConcorrenteTestes(grid, newgrid, N, aux, N, tipo);
                    }
                    arraythreads[j] = new Thread(threads[j]);
                    arraythreads[j].start();
                }
                for (int k = 0; k < num_threads; k++) {
                    try {
                        arraythreads[k].join();
                    } catch (InterruptedException e) {
                        System.out.println("Erro no join");
                        e.printStackTrace();
                    }
                }
                Utilitarios.copiaMatriz(grid, newgrid, N);

                laco_final = System.currentTimeMillis();
                if (i == 0)
                    media_laco = laco_final - laco_inicial;
                media_laco = (media_laco + (laco_final - laco_inicial)) / 2;

//                if (i + 1 == iteracoes)
//                    System.out.println("Geracao " + (i + 1) + ": " + Utilitarios.somaMatriz(grid, N));
            }
            paralelo_final = System.currentTimeMillis();
            total_final = System.currentTimeMillis();

            if (m == 0) {
                media_medias_laco = media_laco;
                media_paralelo = paralelo_final - paralelo_inicial;
                media_total = total_final - total_inicial;
            }

            media_medias_laco = (media_medias_laco + media_laco)/2;
            media_paralelo = (media_paralelo + (paralelo_final - paralelo_inicial))/2;
            media_total = (media_total + (total_final - total_inicial))/2;

            //System.out.println("Media de lacos: " + media_laco + "ms\nTempo total: " + (total_final - total_inicial) + "ms\nTempo em paralelo: " + (paralelo_final - paralelo_inicial) + "ms");
        }
        System.out.println("Media totais: "+media_total+"ms\nMedia paralelos: "+media_paralelo+"ms\nMedia lacos: "+media_medias_laco+"ms");
    }
}
